core/blockchain.go

```diff
@@ -749,11 +795,19 @@ func (self *BlockChain) InsertHeaderChain(chain []*types.Header, checkFreq int)
 			// Verify that the header honors the chain parameters
 			checkPow := verify[index]
 
+			/*
+				var err error
+				if index == 0 {
+					err = self.processor.ValidateHeader(header, checkPow, false)
+				} else {
+					err = self.processor.ValidateHeaderWithParent(header, chain[index-1], checkPow, false)
+				}
+			*/
 			var err error
 			if index == 0 {
-				err = self.processor.ValidateHeader(header, checkPow, false)
+				err = self.validator.ValidateHeader(header, self.GetHeader(header.ParentHash), checkPow)
 			} else {
-				err = self.processor.ValidateHeaderWithParent(header, chain[index-1], checkPow, false)
+				err = self.validator.ValidateHeader(header, chain[index-1], checkPow)
```

Comment:
Since the validator can be set dynamically, these accesses need to be protected by some mutex. Maybe locking the entire `self.mu` in these scenarios is a bit too expensive. Maybe a better approach would be to add a new mutex to self, for setting/getting the validators/processors. Then the above 4-5 getter/setters could also use this, and this code here should be updated a bit so that it obtains a lock before reading the validator. (Note, then it should do this only once at the beginning of the function and not lock/unlock like crazy for each and every header we process iteratively.